### ISSUE TITLE
[Question] Constructing complex constraints

### ISSUE THREAD

**Issue by jonahgeorge** [wouterken/rulp#2]
**tl;dr** I'm curious if there is an undocumented way to better deal with constructing constraints, or if this is an area you would be interested in making the API more flexible.

This question is primarily about constructing constraints but it can also be applied to objective functions. I'm currently in the process of translating a Pulp LP into Rulp to be embedded into a Rails application. In Pulp there is the ability to define a variable as `None` and slowly build the constraint via loops:

``` python
def _students_per_course_constraint(self):
  '''
  This constraint restricts the number of TAs per course. 
  It uses the current_enrollment numbers of the section to determine
  the number of TAs to assign.
  '''
  for section in SECTIONS:
    c = None
    for student in STUDENTS:
      c += self.variables[student["id"]][section["id"]]
    self.problem += (c <= (section["current_enrollment"] // STUDENTS_PER_TEACHING_ASSISTANT))
```

Naively translating this to Ruby you might end up with the code below; however, this results in `method_missing` errors when adding `Rulp::LV`s to the nil constraint. 

``` ruby
def students_per_course_contraint
  SECTIONS.each do |section|
    constraint = nil
    STUDENTS.each do |student|
      constraint += VAR_b(student["id"], section["id"]) 
    end
    @problem[ constraint <= (section["current_enrollment"] / STUDENTS_PER_TA) ]
  end
end
```

As far as I can tell there is no way to do this in Rulp and instead you must construct a complex constraint directly in the `@problem[]` call. For obvious reasons, this code is not nearly as clear about its intentions.

``` ruby
def students_per_course_contraint
  SECTIONS.each do |section|
    @problem[ STUDENTS.map { |student| VAR_b(student["id"], section["id"]) }.inject(:+) <= (section["current_enrollment"] / STUDENTS_PER_TA) ]
  end
end
```

_Working examples of these snippets can be found [here](https://github.com/jonahgeorge/gta-assignment-ilp)._

**Comment by wouterken** [comment]
Hi Jonah

This is definitely possible in Ruby by coercing based on nil or monkey patching NilClass itself. However I'm not too keen to do this as there are many alternate ways to construct these constraints which abides by what some would consider to be more idiomatic ruby.

A port of your Python code which looks almost the same but performs a nil check would look like this

``` ruby
def students_per_course_contraint
  SECTIONS.each do |section|
    constraint = nil
    STUDENTS.each do |student|
      constraint = constraint ? constraint + VAR_b(student["id"], section["id"]) : VAR_b(student["id"], section["id"])
    end
    @problem[ constraint <= (section["current_enrollment"] / STUDENTS_PER_TA) ]
  end
end
```

An alternate way to express the same constraint using reduction

``` ruby
    def students_per_course_contraint
      SECTIONS.each do |section|
        constraint = STUDENTS[1..-1].reduce(VAR_b(STUDENTS[0]['id'], section['id'])) do |memo, student|
          memo + VAR_b(student["id"], section["id"])
        end
        @problem[constraint <= (section["current_enrollment"] / STUDENTS_PER_TA) ]
      end
    end
```

Personally I prefer your initial instinct to map parameters to variables, and construct constraints from this. However you do not have to do this mapping directly within @problem. You can stretch this out a little to have it look quite clean. Something along the lines of:

``` ruby
    def students_per_course_contraint
      SECTIONS.each do |section|
        variables = STUDENTS.map do |student|
          VAR_b(student["id"], section["id"])
        end
        @problem[ variables.sum <= (section["current_enrollment"] / STUDENTS_PER_TA) ]
      end
    end
```

As an aside: I noticed that inside your file you are initialising the LP variables before using them. This step is not actually necessary as the dynamic constant generation will initialise the variables as needed when first used.

**Comment by jonahgeorge** [comment]
I really appreciate your help, it makes a lot more sense to me now. 
Ended up using just a slight variation on your third example:

``` ruby
def students_per_course_contraint
  SECTIONS.each do |section|
    variables = STUDENTS.map { |student| VAR_b(student["id"], section["id"]) }
    constraint = (variables.inject(:+) <= (section["current_enrollment"] / STUDENTS_PER_TA))
    @problem[ constraint ]
  end
end
```

Thank you!
